gsy_framework/community_datasheet/community_datasheet_parser.py

```diff
@@ -0,0 +1,182 @@
+from gsy_framework.community_datasheet.community_datasheet_validator import (
+    CommunityDatasheetValidator)
+from gsy_framework.community_datasheet.exceptions import CommunityDatasheetException
+from gsy_framework.community_datasheet.location_converter import (
+    LocationConverter, LocationConverterException)
+from gsy_framework.constants_limits import FIELDS_REQUIRED_FOR_REBASE
+from gsy_framework.enums import CloudCoverage
+from gsy_framework.scenario_validators import scenario_validator
+
+logger = logging.getLogger(__name__)
+
+PROFILE_KEYS_BY_TYPE = {
+    "Load": "daily_load_profile",
+    "PV": "power_profile",
+    "SmartMeter": "smart_meter_profile"
+}
+
+
+class CommunityDatasheetParser:
+    """Parser for the community datasheet."""
+    def __init__(self, filename):
+        self._datasheet = CommunityDatasheetReader.read(filename)
+
+    def parse(self):
+        """Convert the datasheet to grid-like items."""
+        self._parse_pvs(self._datasheet.pvs, self._datasheet.members)
+        assets_by_member = self._get_assets_by_member()
+
+        self._merge_profiles_into_assets(self._datasheet.profiles, assets_by_member)
+        grid = self._create_grid(assets_by_member)
+
+        self._validate_grid(grid)
+        return {
+            "settings": self._datasheet.settings,
+            "grid": grid
+        }
+
+    def _parse_pvs(self, pvs_by_member: Dict, members_information: Dict):
+        self._add_pv_coordinates(pvs_by_member, members_information)
+
+        pv_assets = (
+            pv_asset for member_assets in pvs_by_member.values() for pv_asset in member_assets)
+        for pv_asset in pv_assets:
+            pv_asset["cloud_coverage"] = self._infer_pv_cloud_coverage(pv_asset)
+
+    def _infer_pv_cloud_coverage(self, pv_asset: Dict) -> int:
+        """Infer which type of profile generation should be used."""
+        asset_name = pv_asset["name"]
+        # The user explicitly provided a profile for the PV
+        if asset_name in self._datasheet.profiles:
+            return CloudCoverage.UPLOAD_PROFILE.value
+
+        # Each PV without profile must provide the parameters needed to call the Rebase API
+        missing_attributes = [
+            field for field in FIELDS_REQUIRED_FOR_REBASE if not pv_asset.get(field)]
+
+        if missing_attributes:
+            raise CommunityDatasheetException(
+                f'The asset "{asset_name}" does not define the following attributes: '
+                f"{missing_attributes}. Either add a profile or provide all the missing fields.")
+
+        return CloudCoverage.LOCAL_GENERATION_PROFILE.value
+
+    @classmethod
+    def _add_pv_coordinates(cls, pvs_by_member: Dict, members_information: Dict):
+        try:
+            location_converter = LocationConverter()
+        except LocationConverterException as ex:
+            logger.warning(ex)
+            return
+
+        with requests.Session() as session:
```

Comment:
In a way, yes. Using sessions should add a `keep-alive` header to each request. If the server supports it (and the geocoder API does), it will open only one TCP connection at the beginning, and then use that for all following requests (rather than open a new TCP connection for each request).

For `requests` docs:
> The Session object allows you to persist certain parameters across requests. It also persists cookies across all requests made from the Session instance, and will use urllib3’s connection pooling. So if you’re making several requests to the same host, the underlying TCP connection will be reused, which can result in a significant performance increase (see HTTP persistent connection).

You can check these docs:
- [Geocoder - Using a Session](https://geocoder.readthedocs.io/api.html#using-a-session)
- [Requests - Session Objects](https://requests.readthedocs.io/en/latest/user/advanced/#session-objects)

Reply:
Thanks a lot for the summary!